### PR TITLE
Refactor 2

### PR DESCRIPTION
**Pull request by mayashavin**
### Brief Summary of Changes
1. Add units test for breakpoints, visibilities.
2. Implement better logic for computing transformation options.
3. Refactor breakpoints code, lazy load.
4. Add logic to find a component based on default name.

#### What does this PR address?
- [ ] Github issue (Add reference - #XX)
- [x] Refactoring
- [ ] New feature
- [x] Bug fix
- [x] Adds more tests

#### Are tests included?
- [x] Yes
- [ ] No

### CODE REVIEW

File: src/helpers/findBreakpoint.js [hunk 1]
```diff
@@ -6,9 +6,7 @@
...
  */
 export function findBreakpoint(stops, value) {
-  return stops
-    .concat([])
-    .sort((a, b) => a - b)
-    .filter(stop => stop >= value)
-    .concat(stop)[0];
+  const filteredBreakpoints = stops.sort((a, b) => a - b)
+                            .filter(stop => stop >= value)
+  return (filteredBreakpoints.length ? filteredBreakpoints : stops)[0]
```
Comment (patrick-tolosa) [inline src/helpers/findBreakpoint.js:11]: can `stops` be something else than an array?

File: tests/unit/helpers/size.spec.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+describe('watchElementSize', () => {
+ const nativeResizeObserver = global.ResizeObserver;
+ afterEach(() => {
+   jest.clearAllMocks()
+   window.ResizeObserver = nativeResizeObserver
+ })
+ it(' there is no resizeObserver', () => {
+   window.ResizeObserver = null
+   // const resizeObserverSpy = jest.spyOn(global, 'ResizeObserver').mockImplementation(null)
```
Comment (patrick-tolosa) [inline tests/unit/helpers/size.spec.js:11]: can we remove the comment?

File: src/mixins/size.js [hunk 3]
```diff
@@ -19,25 +35,27 @@ export const size = {
...
+      if (!this.responsive) {
+        this.cancelSizeListener && this.cancelSizeListener()
+        return
       }
+
+      const isElementRendered = !!this.$el
+
+      if (!isElementRendered || this.cancelSizeListener) return
+
+      this.cancelSizeListener = watchElementSize(this.$el, newSize => {
```
Comment (patrick-tolosa) [inline src/mixins/size.js:47]: I think we're missing a call to `this. cancelSizeListener`, 
We should probably remove the listener at some point

**Review by patrick-tolosa — APPROVED** [review]

File: src/mixins/size.js (hunk 3 above)
Comment (mayashavin) [reply src/mixins/size.js:47]: at `destroyed` :)

File: src/helpers/findBreakpoint.js (hunk 1 above)
Comment (mayashavin) [reply src/helpers/findBreakpoint.js:11]: in JavaScript, anything :P. In general, the only place it was called is in `getResizeTransformation`, and the input is always an Array. But we should put a check here anyway.

**Pull request merged**